db-address-space: init db-address-space crate

HerlinCoder opened the pull request:
This patch introduces db-address-space crate to dragonball-sandbox,
which is designed to support the virtual machine's address space of
guest memory.

Signed-off-by: Alibaba Dragonball Team
Signed-off-by: Helin Guo <helinguo@linux.alibaba.com>

HerlinCoder committed:
db-address-space: init db-address-space crate

This patch introduces db-address-space crate to dragonball-sandbox,
which is designed to support the virtual machine's address space of
guest memory.

Signed-off-by: Alibaba Dragonball Team
Signed-off-by: Helin Guo <helinguo@linux.alibaba.com>

HerlinCoder committed:
address-space: refactor to fix clippy error

we fix some error of cargo clippy

Signed-off-by: Helin Guo <helinguo@linux.alibaba.com>

jiangliu commented on Cargo.toml:
```diff
@@ -3,4 +3,5 @@
 members = [
     "crates/db-allocator",
     "crates/db-arch",
+    "crates/db-address-space"
```
Please keep the list in sorted order.

jiangliu commented on crates/db-address-space/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+[dependencies]
+arc-swap = ">=0.4.8"
+libc = ">=0.2.39"
+nix = "~0.15.0"
```
nix = ">=0.23"?

ZizhengBian replied:
use version directly, instead of `>=` or `~`

studychao commented on crates/db-address-space/src/numa.rs:
```diff
@@ -0,0 +1,66 @@
+// Copyright (C) 2021 Alibaba Cloud. All rights reserved.
```
Could you add some UT for numa.rs?

studychao replied:
could do that in another pull request

studychao committed:
deny: add `MIT` to allowed license

Since some crates we depend are using MIT license, we should allow MIT
licence in deny.toml to pass CI

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

studychao commented on crates/db-address-space/README.md:
```diff
@@ -0,0 +1,41 @@
+
+## Design
+
+db-address-space crate is designed as the address space of guest memory. We define several data structures in this crate to describe logical information abount address space of guest memory for VMM to use.
```
abount -> about

studychao commented on crates/db-address-space/src/numa.rs:
```diff
@@ -0,0 +1,66 @@
+        self.region_infos.push(*info);
+    }
+
+    /// add a new vcpu id belong to this numa node
```
a new vcpu id -> it should be a vec of vcpu ids referred from the code here

studychao commented on crates/db-address-space/src/address_space.rs:
```diff
@@ -0,0 +1,979 @@
+};
+
+/// Internal types for Address space.
+#[cfg(not(feature = "atomic-guest-memory"))]
```
Could you add some description for atomic-guest-memory since it's hard for people to understand its usage

HerlinCoder replied:
I've added some description pls have a look if it is clear

HerlinCoder committed:
address-space: add ut for numa

we add some ut cases for numa in address-space

Signed-off-by: Helin Guo <helinguo@linux.alibaba.com>

jiangliu commented on deny.toml:
```diff
@@ -70,7 +70,7 @@ unlicensed = "deny"
 # [possible values: any SPDX 3.11 short identifier (+ optional exception)].
 allow = [
-    #"MIT",
+    "MIT",
```
which modules is licensed under MIT?

studychao replied:
db-address-space 's dependency `nix`

jiangliu replied:
That's OK then, nix is widely used.

HerlinCoder committed:
address-space: change hotplug to atomic

We change AddressSpaceHotplug to AddressSpaceAtomic so that it is more
readable. This supports atomic address space region, not only suit for
memory hotplug scenario, but also other scenario that add new regions
atomicly.

Signed-off-by: Helin Guo <helinguo@linux.alibaba.com>

jiangliu commented on crates/db-address-space/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "db-address-space"
+version = "0.1.0"
+authors = ["Helin Guo <helinguo@linux.alibaba.com>"]
```
Use "authors = ["Alibaba Dragonball Team"]"?

jiangliu commented on crates/db-address-space/README.md:
```diff
@@ -0,0 +1,41 @@
+
+## Design
+
+db-address-space crate is designed as the address space of guest memory. We define several data structures in this crate to describe logical information about address space of guest memory for VMM to use.
```
The `db-address-space` crate is an address space manager for virtual machines, which manages memory and MMIO resources resident in the guest physical address space.

studychao reviewed (approved):
LGTM. Please remeber to update the README file at dragonabll-sandbox root dir.
